### PR TITLE
feat: Use bidirectional streaming for Source Transformer

### PR DESCRIPTION
**Pull request by BulkBeing**
Related PRs: 
https://github.com/numaproj/numaflow/pull/2071
https://github.com/numaproj/numaflow-rs/pull/91

### CODE REVIEW

**Review by vigith — APPROVED** [review]

File: pkg/sourcetransformer/service.go [hunk 1]
```diff
@@ -34,27 +41,95 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*v1.ReadyResponse,
...
+	if err := stream.Send(handshakeResponse); err != nil {
+		return fmt.Errorf("sending handshake response to client over gRPC stream: %w", err)
+	}
+
+	ctx := stream.Context()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	grp, grpCtx := errgroup.WithContext(ctx)
+
+	senderCh := make(chan *v1.SourceTransformResponse, 500) // TODO: identify the right buffer size
```
Comment (yhl25) [inline pkg/sourcetransformer/service.go:76]: Can you please make it fixme?

File: pkg/sourcetransformer/service.go [hunk 2]
```diff
@@ -34,27 +41,95 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*v1.ReadyResponse,
...
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+		if d.Handshake != nil {
+			return fmt.Errorf("expected source transform messages, received handshake message")
+		}
+
```
Comment (yhl25) [inline pkg/sourcetransformer/service.go:103]: Do we need this check? it should never happen

File: pkg/sourcetransformer/service.go [hunk 3]
```diff
@@ -34,27 +41,95 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*v1.ReadyResponse,
...
 	}
-	responseList := &v1.SourceTransformResponse{
-		Results: results,
+
+	if err := grp.Wait(); err != nil {
+		statusErr := status.Errorf(codes.Internal, err.Error())
+		return statusErr
 	}
-	return responseList, nil
+	return nil
```
Comment (yhl25) [inline pkg/sourcetransformer/service.go:134]: if there is an error we will not let the client know until the stream is closed?

**Review by yhl25 — CHANGES_REQUESTED** [review]
rebase has problems

File: pkg/sourcetransformer/service.go [hunk 4]
```diff
@@ -34,27 +41,95 @@ func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*v1.ReadyResponse,
...
+		if d.Handshake != nil {
+			return fmt.Errorf("expected source transform messages, received handshake message")
+		}
+
+		req := d.Request
+		grp.Go(func() error {
+			var hd = NewHandlerDatum(req.GetValue(), req.EventTime.AsTime(), req.Watermark.AsTime(), req.Headers)
+			messageTs := fs.Transformer.Transform(grpCtx, req.GetKeys(), hd)
+			var results []*v1.SourceTransformResponse_Result
+			for _, m := range messageTs.Items() {
```
Comment (yhl25) [inline pkg/sourcetransformer/service.go:109]: handle panics in the udf

**Comment by BulkBeing** [comment]
> rebase has problems

Fixed by using merge commit instead of rebasing.

File: pkg/sourcetransformer/service_test.go [hunk 5]
```diff
@@ -83,42 +143,120 @@ func TestService_sourceTransformFn(t *testing.T) {
...
+
+			got, err := stream.Recv()
+			assert.NoError(t, err, "Receiving message from the stream")
+
+			assert.Equal(t, got.Results, tt.want.Results)
 		})
 	}
 }
+
+func doHandshake(t *testing.T, stream proto.SourceTransform_SourceTransformFnClient) {
```
Comment (KeranYang) [inline pkg/sourcetransformer/service_test.go:194]: nit: please follow the format to name the unit test. e.g., `TestService_sourceTransformFn`

Comment (BulkBeing) [reply pkg/sourcetransformer/service_test.go:194]: This is a helper function, not a top-level test function.

**Review by yhl25 — APPROVED** [review]

**Pull request merged**